queue/abstract/queue_session.lua

```diff
@@ -0,0 +1,195 @@
+    elseif session_uuid ~= nil then
+        -- Validate UUID.
+        if not pcall(uuid.frombin, session_uuid) then
+            error('Invalid UUID format.')
+        end
+
+        -- identify using a previously created session.
+        -- Check that a session with this uuid exists.
+        local ids_by_uuid = queue_session_ids.index.uuid:select(
+            session_uuid, { limit = 1 })[1]
+        if ids_by_uuid == nil and inactive_sessions[session_uuid] == nil then
+            error('The UUID ' .. uuid.frombin(session_uuid):str() ..
+                ' is unknown.')
+        end
+
+        if cur_uuid ~= session_uuid then
+            if cur_uuid ~= nil then
+                queue_session.remove_connection(conn_id)
+            end
+            queue_session_ids:insert({conn_id, session_uuid})
+            cur_uuid = session_uuid
+        end
+    end
+
+    -- Exclude the session from inactive.
+    inactive_sessions[cur_uuid] = nil
+
+    return cur_uuid
+end
+
+local function on_session_remove(callback)
+    if type(callback) ~= 'function' then
+        error('The "on_session_remove" argument type must be "function"')
+    end
+    queue_session._on_session_remove = callback
+end
+
+--- Remove a connection from the list of active connections and
+-- release its tasks if necessary.
+local function remove_connection(conn_id)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_uuid = queue_session.get_uuid(conn_id)
+
+    queue_session_ids:delete{conn_id}
+    local session_ids = queue_session_ids.index.uuid:select(session_uuid,
+        {limit = 1})[1]
+
+    -- If a queue session doesn't have any active connections it should be
+    -- removed (if ttr is absent) or moved to the "inactive sessions" list.
+    if session_ids == nil then
+        local ttr = queue_session.settings['ttr'] or 0
+        if ttr > 0 then
+            inactive_sessions[session_uuid] = util.event_time(ttr)
+        elseif queue_session._on_session_remove ~= nil then
+            queue_session._on_session_remove(session_uuid)
+        end
+    end
+end
+
+local function set_settings(opts)
+    opts = opts or {}
+
+    for key, val in pairs(opts) do
+        if key == 'ttr' then
+            if val < 0 then
+                error('Invalid value of ttr: ' .. val)
+            end
+        else
+            error('Unknown option ' .. key)
+        end
+    end
+
+    queue_session.settings = opts
+end
+
+local function session_grant(user)
+    box.schema.user.grant(user, 'read, write', 'space', '_queue_session_ids',
+        { if_not_exists = true })
+end
+
+local function start()
+    identification_init()
+    queue_session.expiration_fiber = create_expiration_fiber()
+end
+
+-- methods
+local method = {
+    get_uuid = get_uuid,
+    identify = identify,
+    on_session_remove = on_session_remove,
+    remove_connection = remove_connection,
+    set_settings = set_settings,
+    session_grant = session_grant,
```

Comment:
It gives the ability to magane sessions from given user, right? How about just `grant()`?

Reply:
Accepted.